src/goto-instrument/synthesizer/expr_enumerator.cpp

```diff
@@ -0,0 +1,364 @@
+  // each sub-enumerator to be the corresponding component in the partition.
+  for(const auto &partition : get_partitions(expected_size - 1, arity))
+  {
+    if(!is_good_partition(partition))
+      continue;
+
+    set_sub_enumerator_sizes(partition);
+
+    // Compute the Cartesian product as result.
+    for(const auto &product_tuple : cartesian_product_of_enumerators(
+          sub_enumerators, sub_enumerators.begin()))
+    {
+      // Optimization: rule out equivalent expressions
+      // using certain equivalence class.
+      // Keep only representation tuple of each equivalence class.
+      if(is_equivalence_class_representation(product_tuple))
+        result.insert(simplify_expr(instantiate(product_tuple), ns));
+    }
+  }
+
+  return result;
+}
+
+non_leaf_enumeratort::non_leaf_enumeratort(
+  const enumeratorst &enumerators,
+  const namespacet &ns)
+  : enumerator_baset(ns)
+{
+  /// initialize the list of sub-enumerators
+  sub_enumerators = enumeratorst();
+  for(const auto &g : enumerators)
+  {
+    sub_enumerators.emplace_back(g->clone());
+  }
+
+  /// Arity n is the number of sub-enumerators
+  arity = sub_enumerators.size();
+  INVARIANT(arity > 0, "non-zero arity");
+
+  /// Only enumerate sub-expressions for good size partitions.
+  /// This is an optimization to avoid enumerating equivalent expressions.
+  /// As default: Use no optimization. Accept all partitions.
+  is_good_partition = [](const partitiont &) { return true; };
+}
+
+void non_leaf_enumeratort::set_sub_enumerator_sizes(const partitiont &sizes)
+{
+  INVARIANT(sizes.size() == arity, "arity not equal to number of arguments");
+  auto front = sizes.begin();
+  for(auto &sub_enumerator : sub_enumerators)
+  {
+    sub_enumerator->set_size(*front);
+    std::advance(front, 1);
+  }
+}
+
+std::set<expr_listt> non_leaf_enumeratort::cartesian_product_of_enumerators(
+  enumeratorst &enumerators,
+  const enumeratorst::iterator &it)
+{
+  std::set<expr_listt> result;
+
+  if(std::next(it) == enumerators.end())
+  {
+    /// Current enumerator is the last enumerator.
+    /// Add all expressions enumerated by `it` to `result`.
+    for(const auto &e : enumerators.back()->enumerate())
+    {
+      expr_listt new_tuple;
+      new_tuple.emplace_back(e);
+      result.insert(new_tuple);
+    }
+  }
+  else
+  {
+    /// First compute the Cartesian product of enumerators after `it`.
+    /// And then append the expressions enumerated by the
+    /// `it` to every list in the Cartesian product.
+    for(const auto &sub_tuple :
+        cartesian_product_of_enumerators(enumerators, std::next(it)))
+    {
+      for(const auto &elem : (*it)->enumerate())
+      {
+        expr_listt new_tuple = expr_listt(sub_tuple);
+        new_tuple.emplace_front(elem);
+        result.insert(new_tuple);
+      }
+    }
+  }
+  return result;
+}
+
+std::list<partitiont>
+non_leaf_enumeratort::get_partitions(std::size_t n, std::size_t k)
```

Comment:
So as to avoid having to convince myself that your algorithm is correct: how about using https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation instead?

Reply:
Seconded.  This took quite a bit of time to understand.

Reply:
Yes, I added an implementation that use the bithacks to enumerate bit permutation. However, such method only works for n <= 64 due to the length of size_t. I rewrote the old implementation using a similar idea of next-bit-permutation to handle n > 64

Reply:
You could use `mp_integer`, but is it actually realistic to ever want n > 64? Depending on the value of `k` this might get very expensive...